File: frontend/src/api/pipelines.ts
import { http } from "./http";
import type {
  AnalysisReportDetailResponse,
  AnalysisReportListResponse,
  PipelineActionResponse,
  SystemActionResponse,
  SystemStatusResponse,
  PipelineFrameDetailResponse,
  PipelineFrameRangeResponse,
  PipelineResultResponse,
  PipelineResultSummaryResponse,
  PipelineRunRequest,
  PipelineRunResponse,
  PipelineStatusResponse,
  PipelineTaskListResponse,
} from "../types/video";

export async function runPipeline(body: PipelineRunRequest): Promise<PipelineRunResponse> {
  const { data } = await http.post<PipelineRunResponse>("/api/pipelines/run", body);
  return data;
}

export async function getPipelineStatus(pipelineId: string): Promise<PipelineStatusResponse> {
  const { data } = await http.get<PipelineStatusResponse>(`/api/pipelines/${pipelineId}/status`);
  return data;
}

export async function cancelPipeline(pipelineId: string): Promise<PipelineStatusResponse> {
  const { data } = await http.post<PipelineStatusResponse>(`/api/pipelines/${pipelineId}/cancel`);
  return data;
}

export async function deletePipelineTask(pipelineId: string): Promise<PipelineActionResponse> {
  const { data } = await http.delete<PipelineActionResponse>(`/api/pipelines/${pipelineId}`);
  return data;
}

export async function retryPipeline(pipelineId: string, overwrite = false): Promise<PipelineStatusResponse> {
  const { data } = await http.post<PipelineStatusResponse>(`/api/pipelines/${pipelineId}/retry`, null, {
    params: { overwrite },
  });
  return data;
}

export async function getPipelineResult(pipelineId: string): Promise<PipelineResultResponse> {
  const { data } = await http.get<PipelineResultResponse>(`/api/pipelines/${pipelineId}/result`);
  return data;
}

export async function getPipelineResultSummary(pipelineId: string): Promise<PipelineResultSummaryResponse> {
  const { data } = await http.get<PipelineResultSummaryResponse>(`/api/pipelines/${pipelineId}/result/summary`);
  return data;
}

export async function getPipelineFrameDetail(
  pipelineId: string,
  frame: number,
): Promise<PipelineFrameDetailResponse> {
  const { data } = await http.get<PipelineFrameDetailResponse>(`/api/pipelines/${pipelineId}/result/frame/${frame}`);
  return data;
}

export async function getPipelineFrameRange(
  pipelineId: string,
  params: {
    start_frame?: number;
    end_frame?: number;
    start_sec?: number;
    end_sec?: number;
  },
): Promise<PipelineFrameRangeResponse> {
  const { data } = await http.get<PipelineFrameRangeResponse>(`/api/pipelines/${pipelineId}/result/frames`, { params });
  return data;
}

export async function listPipelineTasks(params?: {
  limit?: number;
  status?: string;
}): Promise<PipelineTaskListResponse> {
  const { data } = await http.get<PipelineTaskListResponse>("/api/pipelines", { params });
  return data;
}


export async function listAnalysisReports(params?: {
  limit?: number;
  q?: string;
  min_score?: number;
  min_confidence?: number;
}): Promise<AnalysisReportListResponse> {
  const { data } = await http.get<AnalysisReportListResponse>("/api/reports", { params });
  return data;
}

export async function getAnalysisReport(pipelineId: string): Promise<AnalysisReportDetailResponse> {
  const { data } = await http.get<AnalysisReportDetailResponse>(`/api/reports/${pipelineId}`);
  return data;
}


const SYSTEM_STATUS_CACHE_TTL_MS = 60_000;
let cachedSystemStatus: SystemStatusResponse | null = null;
let cachedSystemStatusAt = 0;
let systemStatusRequest: Promise<SystemStatusResponse> | null = null;

export async function getSystemStatus(options?: { force?: boolean; maxAgeMs?: number }): Promise<SystemStatusResponse> {
  const maxAgeMs = options?.maxAgeMs ?? SYSTEM_STATUS_CACHE_TTL_MS;
  const now = Date.now();
  if (!options?.force && cachedSystemStatus && now - cachedSystemStatusAt < maxAgeMs) {
    return cachedSystemStatus;
  }
  if (!options?.force && systemStatusRequest) return systemStatusRequest;

  systemStatusRequest = http.get<SystemStatusResponse>("/api/system/status")
    .then(({ data }) => {
      cachedSystemStatus = data;
      cachedSystemStatusAt = Date.now();
      return data;
    })
    .finally(() => {
      systemStatusRequest = null;
    });

  return systemStatusRequest;
}

export function getCachedSystemStatus(): SystemStatusResponse | null {
  return cachedSystemStatus;
}

export async function getSystemStatusUncached(): Promise<SystemStatusResponse> {
  const { data } = await http.get<SystemStatusResponse>("/api/system/status");
  cachedSystemStatus = data;
  cachedSystemStatusAt = Date.now();
  return data;
}

export async function rebuildAnalysisReports(): Promise<SystemActionResponse> {
  const { data } = await http.post<SystemActionResponse>("/api/system/maintenance/rebuild-reports");
  return data;
}


export async function trimOutputPairs(keep: number): Promise<SystemActionResponse> {
  const { data } = await http.post<SystemActionResponse>("/api/system/maintenance/trim-outputs", null, { params: { keep } });
  return data;
}

export async function cleanupDebugArtifacts(): Promise<SystemActionResponse> {
  const { data } = await http.post<SystemActionResponse>("/api/system/maintenance/cleanup-debug");
  return data;
}

File: backend/app/api/pipelines.py
from __future__ import annotations

from fastapi import APIRouter

from app.schemas.dto import (
    PipelineFailureStatsResponse,
    PipelineFrameDetailResponse,
    PipelineFrameRangeResponse,
    PipelineResultResponse,
    PipelineResultSummaryResponse,
    PipelineRunRequest,
    PipelineRunResponse,
    PipelineStatusResponse,
    PipelineTaskListResponse,
)
from app.services.pipeline import (
    get_pipeline_failure_stats,
    get_pipeline_frame_detail,
    get_pipeline_frame_range,
    get_pipeline_result,
    get_pipeline_result_summary,
    get_pipeline_status,
    list_pipeline_tasks,
    request_pipeline_cancel,
    run_pipeline,
)

router = APIRouter(prefix="/pipelines", tags=["pipelines"])


@router.get("", response_model=PipelineTaskListResponse)
async def pipeline_tasks_api(limit: int = 50, status: str | None = None):
    return PipelineTaskListResponse(**list_pipeline_tasks(limit=limit, status=status))


@router.post("/run", response_model=PipelineRunResponse)
async def run_pipeline_api(body: PipelineRunRequest):
    task = run_pipeline(
        teacher_video_id=body.teacher_video_id,
        user_video_id=body.user_video_id,
        overwrite=body.overwrite,
    )
    return PipelineRunResponse(**get_pipeline_status(task.get("pipeline_id", "")))


@router.get("/stats/failures", response_model=PipelineFailureStatsResponse)
async def pipeline_failure_stats_api(days: int = 30, limit: int = 20):
    return PipelineFailureStatsResponse(**get_pipeline_failure_stats(days=days, limit=limit))


@router.get("/{pipeline_id}/status", response_model=PipelineStatusResponse)
async def pipeline_status_api(pipeline_id: str):
    return PipelineStatusResponse(**get_pipeline_status(pipeline_id))


@router.post("/{pipeline_id}/cancel", response_model=PipelineStatusResponse)
async def pipeline_cancel_api(pipeline_id: str):
    return PipelineStatusResponse(**request_pipeline_cancel(pipeline_id))


@router.get("/{pipeline_id}/result", response_model=PipelineResultResponse)
async def pipeline_result_api(pipeline_id: str):
    return PipelineResultResponse(**get_pipeline_result(pipeline_id))


@router.get("/{pipeline_id}/result/summary", response_model=PipelineResultSummaryResponse)
async def pipeline_result_summary_api(pipeline_id: str):
    return PipelineResultSummaryResponse(**get_pipeline_result_summary(pipeline_id))


@router.get("/{pipeline_id}/result/frame/{frame}", response_model=PipelineFrameDetailResponse)
async def pipeline_frame_detail_api(pipeline_id: str, frame: int):
    return PipelineFrameDetailResponse(**get_pipeline_frame_detail(pipeline_id, frame))


@router.get("/{pipeline_id}/result/frames", response_model=PipelineFrameRangeResponse)
async def pipeline_frame_range_api(
    pipeline_id: str,
    start_frame: int | None = None,
    end_frame: int | None = None,
    start_sec: float | None = None,
    end_sec: float | None = None,
):
    return PipelineFrameRangeResponse(
        **get_pipeline_frame_range(
            pipeline_id,
            start_frame=start_frame,
            end_frame=end_frame,
            start_sec=start_sec,
            end_sec=end_sec,
        )
    )
